keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,999 @@
+
+```golang
+type ClusterNetworkPolicy struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec ClusterNetworkPolicySpec
+}
+
+type ClusterNetworkPolicySpec struct {
+	// No implicit isolation of AppliedTo Pods.
+	AppliedTo    AppliedTo
+	Ingress      []ClusterNetworkPolicyIngressRule
+	Egress       []ClusterNetworkPolicyEgressRule
+}
+
+type ClusterNetworkPolicyIngress/EgressRule struct {
+	Action       RuleAction
+	Ports        []networkingv1.NetworkPolicyPort
+	From/To      []ClusterNetworkPolicyPeer
+}
+
+type ClusterNetworkPolicyPeer struct {
+	PodSelector  *metav1.LabelSelector
+	// required if a PodSelector is specified
+	Namespaces   *Namespaces
+	IPBlock      *networkingv1.IPBlock
+}
+
+const (
+  RuleActionAuthorize RuleAction = "Authorize"
+	RuleActionDeny      RuleAction = "Deny"
+	RuleActionAllow     RuleAction = "Allow"
+)
+```
+
+For the ClusterNetworkPolicy ingress/egress rule, the `Action` field dictates whether
+traffic should be allowed or denied or always allowed from/to the ClusterNetworkPolicyPeer.
+This will be a required field.
+
+### ClusterDefaultNetworkPolicy API Design
+The following new `ClusterDefaultNetworkPolicy` API will be added to the `networking.k8s.io` API group:
+
+```golang
+type ClusterDefaultNetworkPolicy struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec ClusterDefaultNetworkPolicySpec
+}
+
+type ClusterDefaultNetworkPolicySpec struct {
+	// Implicit isolation of AppliedTo Pods.
+	AppliedTo   AppliedTo
+	Ingress     []ClusterDefaultNetworkPolicyIngressRule
+	Egress      []ClusterDefaultNetworkPolicyEgressRule
+}
+
+type ClusterDefaultNetworkPolicyIngress/EgressRule struct {
+	Ports            []networkingv1.NetworkPolicyPort
+	OnlyFrom/OnlyTo  []ClusterDefaultNetworkPolicyPeer
+}
+
+type ClusterDefaultNetworkPolicyPeer struct {
+	PodSelector  *metav1.LabelSelector
+	// required if a PodSelector is specified
+	Namespaces   *Namespaces
+	IPBlock      *networkingv1.IPBlock
+}
+```
+
+Most structs above are very similar to NetworkPolicy and quite self-explanatory.
+One detail to notice is that in the ClusterDefaultNetworkPolicy Ingress/Egress rule spec,
+the peers are created in a field named `OnlyFrom`/`OnlyTo`, as opposed to `To`/`From`
+in ClusterNetworkPolicy. We chose this naming to better hint policy writers about
+the isolation effect of ClusterDefaultNetworkPolicy on the workloads it applies to.
+
+### Shared API Design
+The following structs will be added to the `networking.k8s.io` API group and
+shared between `ClusterNetworkPolicy` and `ClusterDefaultNetworkPolicy`:
+
+```golang
+type AppliedTo struct {
+	// required if a PodSelector is specified
+	NamespaceSelector   *metav1.LabelSelector
+	// optional
+	PodSelector         *metav1.LabelSelector
+}
+
+type Namespaces struct {
+	// Self and Selector are mutually-exclusive
+	Self       bool
+	Selector   *metav1.LabelSelector
+}
+```
+
+#### AppliedTo
+The `AppliedTo` field replaces `PodSelector` in NetworkPolicy spec, as means to specify
```

Comment:
> The `AppliedTo` field replaces `PodSelector` in NetworkPolicy spec,

Is this suggesting a change to the existing NetworkPolicy spec? The wording sounds like it, but my understanding is that we're not intending to touch the existing API as part of this.

Reply:
Will re-word